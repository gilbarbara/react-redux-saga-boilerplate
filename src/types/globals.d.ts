declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__: any;
    HIDE_LOGS: boolean;
    store: any;
  }

  const VERSION: string;
}

export {};
