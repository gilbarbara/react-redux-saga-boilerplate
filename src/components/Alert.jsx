import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import theme from 'modules/theme';

import { Alert as AlertComponent, Box, utils } from 'styled-minimal';
import Icon from 'components/Icon';

const { colors, palette } = utils.themeGet(theme);
const variants = { ...colors, ...palette };

AlertComponent.displayName = 'AlertComponent';

const StyledAlert = styled(AlertComponent)`
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
`;

const AlertIcon = styled(Icon)`
  align-items: flex-start;
  background-color: ${({ variant }) => variants[variant]};
  display: flex;
  padding: ${utils.spacer(3)};
`;

const AlertButton = styled.button`
  background-color: ${({ variant }) => variants[variant]};
  pointer-events: all;
  position: absolute;
  right: ${utils.spacer(1)};
  top: ${utils.spacer(1)};
`;

const Alert = ({ children, handleClickClose, id, icon, ...rest }) => {
  const output = {};
  let name;

  switch (rest.variant) {
    case 'success': {
      name = icon || 'check-circle';
      break;
    }
    case 'warning': {
      name = icon || 'exclamation-circle';
      break;
    }
    case 'danger': {
      name = icon || 'times-circle';
      break;
    }
    case 'info': {
      name = icon || 'question-circle';
      break;
    }
    case 'dark': {
      name = icon || 'bell-o';
      break;
    }
    default: {
      name = icon || 'dot-circle-o';
    }
  }

  if (handleClickClose) {
    output.button = (
      <AlertButton data-id={id} onClick={handleClickClose} type="button">
        <Icon name="times" color="#ccc" width={10} />
      </AlertButton>
    );
  }

  return (
    <StyledAlert {...rest}>
      <AlertIcon {...rest} name={name} color="#fff" width={24} />
      <Box p={2} pr={handleClickClose ? 3 : 2}>
        {children}
      </Box>
      {output.button}
    </StyledAlert>
  );
};

Alert.propTypes = {
  children: PropTypes.node.isRequired,
  handleClickClose: PropTypes.func,
  icon: PropTypes.string,
  id: PropTypes.string,
  outline: PropTypes.bool,
  variant: PropTypes.string,
};

Alert.defaultProps = {
  outline: true,
  variant: 'info',
};

export default Alert;
