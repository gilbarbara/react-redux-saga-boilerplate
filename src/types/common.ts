import { ComponentProps, ComponentType } from 'react';
import { Alert, Types } from '@gilbarbara/components';

export type AlertPosition = 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
export type AlertType = ComponentProps<typeof Alert>['type'];

export interface AsyncFlow {
  message: string;
  status: Status;
}

export type GenericFunction<T = any> = (...arguments_: any[]) => T;

export type PlainObject<T = any> = Record<string, T>;

export interface RouteProps {
  component: ComponentType<any>;
  exact?: boolean;
  isAuthenticated: boolean;
  path: string;
  to?: string;
}

export interface Status {
  ERROR: 'error';
  IDLE: 'idle';
  READY: 'ready';
  RUNNING: 'running';
  SUCCESS: 'success';
}

export interface SetAppOOptions {
  query: string;
}

export interface ShowAlertOptions {
  icon?: Types.Icons;
  id?: string;
  position?: AlertPosition;
  timeout?: number;
  type?: AlertType;
}

export type Transitions = 'fade' | 'slideDown' | 'slideLeft' | 'slideRight' | 'slideUp';
