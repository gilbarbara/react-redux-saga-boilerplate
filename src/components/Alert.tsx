import { MouseEvent, ReactNode } from 'react';
import styled from '@emotion/styled';
import { Alert as AlertComponent, Box, Icon, Types } from '@gilbarbara/components';

import theme from '~/modules/theme';

import { AlertType } from '~/types';

interface Props {
  children: ReactNode;
  handleClickClose?: (event: MouseEvent<HTMLButtonElement>) => void;
  icon?: Types.Icons;
  id?: string;
  type?: AlertType;
}

AlertComponent.displayName = 'AlertComponent';

const AlertButton = styled.button`
  color: #fff;
  line-height: 0;
  pointer-events: all;
  position: absolute;
  right: ${theme.spacing.xs};
  top: ${theme.spacing.xs};
`;

function Alert({ children, handleClickClose, id, type = 'neutral', ...rest }: Props) {
  const output: Record<string, any> = {};

  if (handleClickClose) {
    output.button = (
      <AlertButton data-id={id} onClick={handleClickClose} type="button">
        <Icon name="times" size={10} />
      </AlertButton>
    );
  }

  return (
    <AlertComponent {...rest} align="center" data-component-name="Alert" type={type}>
      <Box pl={handleClickClose ? 'sm' : 'xs'}>{children}</Box>
      {output.button}
    </AlertComponent>
  );
}

export default Alert;
