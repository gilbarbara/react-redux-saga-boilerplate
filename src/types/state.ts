import { ReactNode } from 'react';
import { Types } from '@gilbarbara/components';
import { Dispatch } from 'redux';
import { ValueOf } from 'type-fest';

import { AlertPosition, AlertType, Status } from './common';

export interface AlertData {
  icon: Types.Icons;
  id: string;
  message: ReactNode;
  position: AlertPosition;
  timeout: number;
  type: AlertType;
}

export interface Topic {
  cached: boolean;
  data: Array<Record<string, any>>;
  message: string;
  status: ValueOf<Status>;
  updatedAt: number;
}

export interface AlertsState {
  data: AlertData[];
}

export interface AppState {
  query: string;
}

export interface GitHubState {
  topics: Record<string, Topic>;
}

export interface UserState {
  isAuthenticated: boolean;
  status: ValueOf<Status>;
}

export interface RootState {
  alerts: AlertsState;
  app: AppState;
  github: GitHubState;
  user: UserState;
}

export interface WithDispatch {
  dispatch: Dispatch;
}
