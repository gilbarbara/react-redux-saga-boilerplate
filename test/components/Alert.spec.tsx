import React, { ReactNode } from 'react';
import { render, screen } from 'test-utils';

import Alert from '~/components/Alert';

import { AlertType } from '~/types';

describe('Alert', () => {
  it('should render the default alert', () => {
    render(<Alert>Basic Alert</Alert>);

    expect(screen.getByRole('alert')).toMatchSnapshot();
  });

  it.each<{ children: ReactNode; type: AlertType }>([
    { type: 'success', children: 'This is a success message' },
    { type: 'error', children: <p key="danger">This is an error message</p> },
    { type: 'warning', children: 'This is a warning message' },
    { type: 'info', children: 'This is an info message' },
    { type: 'neutral', children: <div key="dark">This is a message</div> },
  ])('should render the $type type', ({ children, type }) => {
    render(<Alert type={type}>{children}</Alert>);

    expect(screen.getByRole('alert')).toMatchSnapshot();
  });
});
