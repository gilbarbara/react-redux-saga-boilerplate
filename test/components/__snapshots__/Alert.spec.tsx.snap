// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert > should render the 'error' type 1`] = `
<div
  class="css-waw36"
  data-component-name="Alert"
  role="alert"
>
  <span
    class="css-nri15g"
    data-component-name="Icon"
    name="close-circle"
  >
    <svg
      height="20"
      src="https://files.gilbarbara.dev/icons/close-circle.svg"
      title="close-circle"
      width="20"
    />
  </span>
  <span
    class="css-lo0v0o"
    data-component-name="Text"
  >
    <div
      class="css-n1d5o7"
      data-component-name="Box"
    >
      <p>
        This is an error message
      </p>
    </div>
  </span>
</div>
`;

exports[`Alert > should render the 'info' type 1`] = `
<div
  class="css-17nincq"
  data-component-name="Alert"
  role="alert"
>
  <span
    class="css-nri15g"
    data-component-name="Icon"
    name="info-circle"
  >
    <svg
      height="20"
      src="https://files.gilbarbara.dev/icons/info-circle.svg"
      title="info-circle"
      width="20"
    />
  </span>
  <span
    class="css-lo0v0o"
    data-component-name="Text"
  >
    <div
      class="css-n1d5o7"
      data-component-name="Box"
    >
      This is an info message
    </div>
  </span>
</div>
`;

exports[`Alert > should render the 'neutral' type 1`] = `
<div
  class="css-1wzqtgx"
  data-component-name="Alert"
  role="alert"
>
  <span
    class="css-nri15g"
    data-component-name="Icon"
    name="message-circle"
  >
    <svg
      height="20"
      src="https://files.gilbarbara.dev/icons/message-circle.svg"
      title="message-circle"
      width="20"
    />
  </span>
  <span
    class="css-lo0v0o"
    data-component-name="Text"
  >
    <div
      class="css-n1d5o7"
      data-component-name="Box"
    >
      <div>
        This is a message
      </div>
    </div>
  </span>
</div>
`;

exports[`Alert > should render the 'success' type 1`] = `
<div
  class="css-d735a"
  data-component-name="Alert"
  role="alert"
>
  <span
    class="css-nri15g"
    data-component-name="Icon"
    name="check-circle"
  >
    <svg
      height="20"
      src="https://files.gilbarbara.dev/icons/check-circle.svg"
      title="check-circle"
      width="20"
    />
  </span>
  <span
    class="css-lo0v0o"
    data-component-name="Text"
  >
    <div
      class="css-n1d5o7"
      data-component-name="Box"
    >
      This is a success message
    </div>
  </span>
</div>
`;

exports[`Alert > should render the 'warning' type 1`] = `
<div
  class="css-17sh1w2"
  data-component-name="Alert"
  role="alert"
>
  <span
    class="css-nri15g"
    data-component-name="Icon"
    name="danger-circle"
  >
    <svg
      height="20"
      src="https://files.gilbarbara.dev/icons/danger-circle.svg"
      title="danger-circle"
      width="20"
    />
  </span>
  <span
    class="css-lo0v0o"
    data-component-name="Text"
  >
    <div
      class="css-n1d5o7"
      data-component-name="Box"
    >
      This is a warning message
    </div>
  </span>
</div>
`;

exports[`Alert > should render the default alert 1`] = `
<div
  class="css-1wzqtgx"
  data-component-name="Alert"
  role="alert"
>
  <span
    class="css-nri15g"
    data-component-name="Icon"
    name="message-circle"
  >
    <svg
      height="20"
      src="https://files.gilbarbara.dev/icons/message-circle.svg"
      title="message-circle"
      width="20"
    />
  </span>
  <span
    class="css-lo0v0o"
    data-component-name="Text"
  >
    <div
      class="css-n1d5o7"
      data-component-name="Box"
    >
      Basic Alert
    </div>
  </span>
</div>
`;
